File: 백준/Silver/28066. 타노스는 요세푸스가 밉다/타노스는 요세푸스가 밉다.py
import sys
from collections import deque

input = sys.stdin.readline

n, k = map(int, input().split())

queue = deque(range(1, n + 1))

while len(queue) > 1:
    # 1마리 남을 때까지 제거
    if len(queue) <= k:
        # 현재 n마리가 k마리보다 적으면 첫번쩨 제외 모든 청설모 제거 => 첫번째가 답이 됨
        print(queue[0])
        break

    # 시계방향으로 현재 것은 뒤로 넘기기
    queue.append(queue.popleft())  # 첫번째 이후
    # 2~k번째들 제거 (k-1)마리 제거
    for _ in range(k - 1):
        queue.popleft()

File: 백준/Silver/28066. 타노스는 요세푸스가 밉다/타노스는 요세푸스가 밉다.kt
package org.example

fun main() {
    val br = System.`in`.bufferedReader()
    val (n, k) = br.readLine().split(" ").map { it.toInt() }
    val queue = ArrayDeque((1..n).toList())

    while (queue.size > 1) {
        if (queue.size <= k) {
            println(queue.first())
            break
        }
        queue.addLast(queue.removeFirst())
        repeat(k - 1) {
            queue.removeFirst()
        }
    }

}